### PR TITLE
Power Code Cleanup + Conversion to Regional Powernets [Refactor Only]

### PR DESCRIPTION
**Pull request by Sirryan2002**
## What Does This PR Do
2nd attempt at Refactors and thorough documentation of the following:

- `/datum/powernet` - converts it into `/datum/regional_powernet` and documents/improves code
- cable - bringing code up to date and improving code
- `/machinery/power` - moves some of its procs/code off of this type and offloads the heavy lifting onto powernets
- engines - engine code needs to be brought up to standards code-wise and should be better documented
- SSmachines way of handling powernets

Also finally kills `/obj/machinery/logic` you will not be missed

## Why It's Good For The Game
Higher quality documented code opens the doors (and makes it easier) for stuff like #19786 and #19734

## Testing
Compiled, ran on a test server

- Started Round
- APCs are charging properly
- Laid wires and cut wires
- Checked SMES to make sure they were charging and outputting power

**I STILL NEED TO TEST SOME ENGINES**

No changelog as there are no player-facing changes

### CODE REVIEW

File: paradise.dme [hunk 1]
```diff
@@ -2298,26 +2299,48 @@
...
+#include "code\modules\power\generators\tracker.dm"
+#include "code\modules\power\generators\treadmill.dm"
+#include "code\modules\power\generators\turbine.dm"
+#include "code\modules\power\generators\portable generators\pacman.dm"
+#include "code\modules\power\generators\portable generators\port_gen.dm"
 #include "code\modules\power\powernets\local_powernet.dm"
+<<<<<<< HEAD
+#include "code\modules\power\powernets\powernet_helpers.dm"
+#include "code\modules\power\powernets\regional_powernet.dm"
+=======
```
Comment (GDNgit) [inline paradise.dme:2343]: leftover from a merge conflict I assume

File: code/modules/power/cables/cable.dm [hunk 2]
```diff
@@ -0,0 +1,449 @@
...
+/*
+	* # /obj/structure/cable
+	*
+	* The red wire thingies you see on the ground all over the station in maintenance
+	* the d1 and d2 vars deal with the "directions" of the cables, since all instances of this cable structure are
+	* just lines, they have two endpoints (d1 and d2).
+*/
+/obj/structure/cable
+	name = "power cable"
+	desc = "A flexible superconducting cable for heavy-duty power transfer"
```
Comment (Henri215) [inline code/modules/power/cables/cable.dm:24]:
```suggestion
	desc = "A flexible superconducting cable for heavy-duty power transfer."
```

File: code/modules/power/generators/portable generators/port_gen.dm [hunk 3]
```diff
@@ -0,0 +1,86 @@
+
+
+//Baseline portable generator. Has all the default handling. Not intended to be used on it's own (since it generates unlimited power).
+/obj/machinery/power/port_gen
+	name = "Placeholder Generator"	//seriously, don't use this. It can't be anchored without VV magic.
+	desc = "A portable generator for emergency backup power"
```
Comment (Henri215) [inline code/modules/power/generators/portable generators/port_gen.dm:6]:
```suggestion
	desc = "A portable generator for emergency backup power."
```

**Comment by Henri215** [comment]
image
I could place cables under glass floors on the direction my character was facing, allowing me to do it once per floor before receiving the "You can't interact with something that's under the floor!" warning.
As you can see in the bottom i managed to replace cables and even get shocked.

File: code/game/turfs/turf.dm [hunk 4]
```diff
@@ -468,7 +469,32 @@
 	return TRUE
 
 /turf/proc/can_lay_cable()
-	return can_have_cabling() & !intact
+	return can_have_cabling() && !intact
+
+/*
+	* # power_list()
+	* returns a list power machinery on the turf and cables on the truf that have a direction equal to the one supplied in params
```
Comment (Contrabang) [inline code/game/turfs/turf.dm:476]:
```suggestion
	* returns a list power machinery on the turf and cables on the turf that have a direction equal to the one supplied in params
```

File: code/modules/power/generators/portable generators/port_gen.dm [hunk 5]
```diff
@@ -0,0 +1,86 @@
+
+
+//Baseline portable generator. Has all the default handling. Not intended to be used on it's own (since it generates unlimited power).
+/obj/machinery/power/port_gen
+	name = "Placeholder Generator"	//seriously, don't use this. It can't be anchored without VV magic.
+	desc = "A portable generator for emergency backup power."
```
Comment (Contrabang) [inline code/modules/power/generators/portable generators/port_gen.dm:6]: Perhaps a warning should be placed in the description/name/something similar that this is not intended for IC use

File: code/modules/power/generators/portable generators/port_gen.dm [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+	var/power_gen = 5000
+	var/power_output = 1
+	var/base_icon = "portgen0"
+
+/obj/machinery/power/port_gen/attack_hand(mob/user as mob)
+	if(..())
+		return
+	if(!anchored)
+		return
+
```
Comment (Contrabang) [inline code/modules/power/generators/portable generators/port_gen.dm:22]: Does this attack_hand here actually do anything?

File: code/__DEFINES/directions.dm [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+//#define SOUTHEAST 6
+//#define NORTHWEST 9
+//#define SOUTHWEST 10
+
+
+
+///Flips a direction along the horizontal axis, will convert E -> W, W -> E, NE -> NW, SE -> SW, etc
+#define FLIP_DIR_HORIZONTALLY(dir) (dir ^ (EAST|WEST))
+///Flips a direction along the vertical axis, will convert N -> S, S -> N, NE -> SE, SW -> NW, etc
+#define FLIP_DIR_VERTICALLY(dir) (dir ^ (NORTH|SOUTH))
```
Comment (Contrabang) [inline code/__DEFINES/directions.dm:18]: Some explanation on how these work would be nice, similar to IS_DIR_DIAGONAL

File: code/modules/power/power_machine.dm [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+///////////////////////////////////////////////
+
+
+// return a knot cable (O-X) if one is present in the turf
+// null if there's none
+/turf/proc/get_cable_node()
+	if(!can_have_cabling())
+		return null
+	for(var/obj/structure/cable/C in src)
+		if(C.d1 == 0)
```
Comment (Contrabang) [inline code/modules/power/power_machine.dm:97]:
```suggestion
		if(C.d1 == NO_DIRECTION)
```

File: code/game/mecha/equipment/tools/work_tools.dm [hunk 9]
```diff
@@ -438,7 +438,7 @@
 	NC.d2 = fdirn
```
Comment (Contrabang) [inline code/game/mecha/equipment/tools/work_tools.dm:438]: Could you change line 437 here to use NO_DIRECTION

**Comment by ppi13** [comment]
> image I could place cables under glass floors on the direction my character was facing, allowing me to do it once per floor before receiving the "You can't interact with something that's under the floor!" warning. As you can see in the bottom i managed to replace cables and even get shocked.

This is unrelated to this pr, and you can do this on live

Both the SM, Tesla and Singularity works fine

However, if you cut wires, then uh, both
image
image
they are still connected even after cutting the wire.

File: code/modules/power/generators/portable generators/port_gen.dm (hunk 6 above)
Comment (Sirryan2002) [reply code/modules/power/generators/portable generators/port_gen.dm]: no, I removed it

File: code/__DEFINES/directions.dm (hunk 7 above)
Comment (Sirryan2002) [reply code/__DEFINES/directions.dm:21]: done

File: code/game/mecha/equipment/tools/work_tools.dm (hunk 9 above)
Comment (Sirryan2002) [reply code/game/mecha/equipment/tools/work_tools.dm:438]: added

**Comment by Sirryan2002** [comment]
> > image I could place cables under glass floors on the direction my character was facing, allowing me to do it once per floor before receiving the "You can't interact with something that's under the floor!" warning. As you can see in the bottom i managed to replace cables and even get shocked.
> 
> This is unrelated to this pr, and you can do this on live
> 
> Both the SM, Tesla and Singularity works fine
> 
> However, if you cut wires, then uh, both image image they are still connected even after cutting the wire.

both should be fixed now in latest commit

**Comment by Inferno1882** [comment]
This seems like a _new_ issue so I assume it is connected to the PR. 

It would appear that all round-start station power cable being fed by an SMES is not receiving any power at all. Cutting and restoring the cable appears to allow the SMES to send power again.

Additionally, APCs are not showing on any Power Monitoring Console until you disconnect and reconnect them to the grid.

Mid comment writing edit: It broke again while testing on live.

tl;dr The code is not playing nice but the APCs don't really seem to care about that.

**Comment by SteelSlayer** [comment]
Inferno's comment pretty much summed up the issues I was seeing when just observing a separate round. It seems that APCs were simply not being registered as a power consumer round-start. It also seems that the issue does not always occur. It also didn't happen in my local testing. 

Round ID: 35673, Map: Cyberiad.
There were no related runtimes that I could see.

Pulling the TM for now.

**Comment by Sirryan2002** [comment]
ill look into it and see if I can get it fixed this weekend

**Comment by Sirryan2002** [comment]
Unable to reproduce this locally, booted up a test server 3 time without seeing this issue. Can you confirm that this is not something you observed before the round fully started (i.g. after "round-start")?

**Comment by Contrabang** [comment]
I was the CE the round that happened on the cyberiad. I can confirm that it was going on the entire round. The only way to get the APCs to work would be to cut the wire under their terminal, then reconnect the terminal to the main grid.

**Comment by Sirryan2002** [comment]
all issues with APCs not charging and the power monitoring computer have been fixed in the latest commits

**Review by SteelSlayer — APPROVED** [review]
Looks good and has been TM'd for a week without any further issues.

Just one spelling nitpick/question.

File: code/modules/power/powernets/README.md [hunk 10]
```diff
@@ -0,0 +1,31 @@
+
+# Understanding Powernets
+Much like any other massive numbers system in SS13, the power (or powernet) system is complex and confusing to work with, only being trumped in complexity by atmospherics/LINDA. This README serves as a powernets 101 guide and breaks down how the system works.
+
+## Two Types of Powernets
+There are two types of powernets in our code
+1. Regional Powernets
+2. Local Powernets
+
+They are two completely different datum types from eachother and serve different completely different purposes. In a nutshell, regional powernets are dynamically sized and deal with physical machinery, cables, and generators whereas local powernets area statically locked into a single area each and work directly with APCs to handle individual machines interactions with the larger regional powernet.
```
[the remark this answers is not in the document]
Comment (Sirryan2002) [reply code/modules/power/powernets/README.md]: yes it's correct, I don't touch local powernets in this PR just reference them in the readme
